scripts/genererate_drivers.py

```diff
@@ -0,0 +1,633 @@
+    :return: a string which is the .h for file the device driver
+    """
+    template = string.Template(textwrap.dedent(METAL_DEV_HDR_TMPL))
+
+    return template.substitute(
+        vendor=vendor,
+        device=device,
+        cap_device=device.upper(),
+        index=str(index),
+        base_address=hex(base_address),
+        vtable=generate_vtable_declarations(device, reglist),
+        metal_device=generate_metal_vtable_definition(device),
+        protos=generate_protos(device, reglist)
+    )
+
+
+# the template for the driver .c file
+METAL_DEV_DRV_TMPL = \
+    """
+    #include <stdint.h>
+    #include <stdlib.h>
+    
+    #include <${device}/${vendor}_${device}${index}.h>
+    #include <metal/compiler.h>
+    #include <metal/io.h>
+    
+    ${base_functions}
+    
+    ${metal_functions}
+    
+    __METAL_DEFINE_VTABLE(metal_${device}) = {
+    ${def_vtable}
+    };
+    
+    const struct metal_${device}* ${device}_tables[] = {&metal_${device}};
+    uint8_t ${device}_tables_cnt = 1;
+    
+    const struct metal_${device}* get_metal_${device}(uint8_t idx)
+    {
+        if (idx >= ${device}_tables_cnt)
+            return NULL;
+        return ${device}_tables[idx];
+    }
+    """
+
+
+def generate_def_vtable(device: str, reg_list: t.List[Register]) -> str:
+    """
+    Generate vtable settings for vtable declaration in .c file
+
+    :param device: the name of the device
+    :param reg_list: the register list for the device
+    :return: the declarations in the vtable for the driver .c file
+    """
+    rv: t.List[str] = []
+    cap_device = device.upper()
+    head = f'    .{device}_base = (uint32_t *){cap_device}_BASE,'
+    rv.append(head)
+    for a_reg in reg_list:
+        reg_name = a_reg.name.lower()
+
+        write_func = f'    .vtable.v_{device}_{reg_name}_write ' \
+                     f'= {device}_{reg_name}_write,'
+        read_func = f'    .vtable.v_{device}_{reg_name}_read ' \
+                    f'= {device}_{reg_name}_read,'
+        rv.append(write_func)
+        rv.append(read_func)
+
+    return '\n'.join(rv)
+
+
+def generate_base_functions(device: str, reg_list: t.List[Register]) -> str:
+    """
+    Generates the basic, not exported register access functions for
+    a given device and register list.
+
+    :param device: the name of the device
+    :param reg_list: the list of registers for the device.
+    :return:  the c code for the register access functions
+    """
+    cap_device = device.upper()
+    rv: t.List[str] = []
+
+    for a_reg in reg_list:
+        name = a_reg.name.lower()
+        cap_name = a_reg.name.upper()
+        size = a_reg.width
+
+        write_func = f"""
+            void {device}_{name}_write(uint32_t *{device}_base, uint{size}_t data)
+            {{
+                volatile uint32_t *control_base = {device}_base;
+                METAL_{cap_device}_REGW(METAL_{cap_device}_{cap_name}) = data;
+            }}"""
```

Comment:
Be careful about newlines. I think you should put the closing triple quotes on their own line here, since that'll ensure that there is a trailing newline that facilitates concatenating the string with other strings, and it will at least be consistent with the `read_func` defined below.

Reply:
fixed